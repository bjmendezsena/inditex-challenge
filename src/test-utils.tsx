import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { theme } from "./theme";

const DefaultTheme = extendTheme(theme);

export const render = (ui: JSX.Element, theme = DefaultTheme): RenderResult => {
  const wrapper = ({ children }: { children?: React.ReactNode }) => (
    <ChakraProvider theme={theme}>{children}</ChakraProvider>
  );

  return testingLibraryRender(ui, { wrapper });
};
