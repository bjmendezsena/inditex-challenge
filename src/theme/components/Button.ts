import type { ButtonThemeConfiguration } from "../../components/Button";
import { buttonVariants, buttonSizes } from "./utils";

export const ConfigButton: ButtonThemeConfiguration = {
  baseStyle: {
    minHeight: 12,
    height: "unset",
    minWidth: 19,
    paddingY: 2,
    paddingX: 4,
    justifyContent: "center",
    alignItems: "center",
    apply: "textStyles.ElementsMediumSM",
    borderRadius: 8,
    transition: "all 0.2s",
    display: "inline-flex",
    color: "grey.400",
    _disabled: {
      opacity: 1,
      backgroundColor: "grey.200",
      color: "grey.400",
    },
    _loading: {
      backgroundColor: "grey.200",
      position: "relative",
      color: "white",
      zIndex: 0,
      _hover: {
        backgroundColor: "grey.200",
      },
    },
  },
  sizes: buttonSizes,
  variants: buttonVariants,
  defaultProps: {
    size: "lg",
    variant: "primary",
  },
};
