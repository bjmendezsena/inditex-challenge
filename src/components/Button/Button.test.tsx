import React from "react";
import { screen } from "@testing-library/react";
import { render } from "../../test-utils";
import type { ButtonProps } from "./Button";
import { Button } from "./Button";

describe(`<Button />`, () => {
  const defaultText = "Content of the button";
  const defaultProps: ButtonProps = {
    children: <>{defaultText}</>,
  };

  const baseSizesList = ["md", "lg"];
  const baseVariantsList = ["primary", "tertiary", "secondary"];
  const statusColors = ["success", "error", "warning", "info"];
  const factoryComponent = (props: ButtonProps = defaultProps) =>
    render(<Button {...props} />);

  test("renders correctly", () => {
    const view = factoryComponent();

    expect(view.asFragment()).toMatchSnapshot();
  });

  test("shows correctly children", () => {
    const label = "Text of the button";
    factoryComponent({ children: <>{label}</> });

    expect(screen.getByText(label)).toBeVisible();
  });

  test("pass down props", () => {
    factoryComponent({
      isDisabled: true,
      children: <>{defaultText}</>,
    });

    expect(screen.getByText(defaultText)).toBeDisabled();
  });

  baseVariantsList.forEach((variant) => {
    baseSizesList.forEach((size) => {
      statusColors.forEach((status) => {
        test(`renders correctly for size ${size} & variant ${variant}`, () => {
          const view = factoryComponent({
            ...defaultProps,
            size,
            variant,
            status,
          });
          expect(view.asFragment()).toMatchSnapshot(
            `Button-${variant}-${size}-${status}`
          );
        });
      });
    });
  });
});
