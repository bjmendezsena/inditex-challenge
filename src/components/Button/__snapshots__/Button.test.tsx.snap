// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders correctly 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-rajj9k"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant primary: Button-primary-lg-error 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-jymy6g"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant primary: Button-primary-lg-info 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-ett4rg"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant primary: Button-primary-lg-success 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-e3rcup"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant primary: Button-primary-lg-warning 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-wj1ala"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant secondary: Button-secondary-lg-error 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-1kx1fhq"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant secondary: Button-secondary-lg-info 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-9d1pl2"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant secondary: Button-secondary-lg-success 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-178a8wn"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant secondary: Button-secondary-lg-warning 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-6hlnjf"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant tertiary: Button-tertiary-lg-error 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-8lffqo"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant tertiary: Button-tertiary-lg-info 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-1svmljl"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant tertiary: Button-tertiary-lg-success 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-m9fpsm"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size lg & variant tertiary: Button-tertiary-lg-warning 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-qew0cg"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant primary: Button-primary-md-error 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-cgs9di"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant primary: Button-primary-md-info 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-10gbpej"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant primary: Button-primary-md-success 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-fum8yl"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant primary: Button-primary-md-warning 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-jnv93u"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant secondary: Button-secondary-md-error 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-4s23xp"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant secondary: Button-secondary-md-info 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-16w97fb"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant secondary: Button-secondary-md-success 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-1ysgpzs"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant secondary: Button-secondary-md-warning 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-1ds3cjo"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant tertiary: Button-tertiary-md-error 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-1zus9n"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant tertiary: Button-tertiary-md-info 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-wuhl4a"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant tertiary: Button-tertiary-md-success 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-29iwb9"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Button /> renders correctly for size md & variant tertiary: Button-tertiary-md-warning 1`] = `
<DocumentFragment>
  <button
    class="chakra-button css-kasexb"
    type="button"
  >
    Content of the button
  </button>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
